### PR TITLE
Preview spectral decon result

### PR DESCRIPTION
**Pull request by ansgarkorf**
Updated GC spectral deconvolution module. 
- added 2 different algorithms
- introduced option to ignore mz values as representative feature (see example 173 in screenshot)
image

- Interactive preview shows clustered features as compound in same color
- Grey marker shows representative feature rt range 
image

- Orange shows retention time tolerance parameter 
image

### CODE REVIEW

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java [hunk 1]
```diff
@@ -166,6 +134,52 @@ private List<FeatureListRow> generatePseudoSpectraByRtOnly(List<ModularFeature>
     return deconvolutedFeatureListRowsByRtOnly;
   }
 
+  private ModularFeature getMainFeature(List<ModularFeature> groups) {
+    List<Range<Double>> adjustedRanges = new ArrayList<>();
```
Comment (SteffenHeu) [inline mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java:138]: the ranges need to be adjusted only once, or? then extract this to another method and call outside of the loop. could also be a RangeMap, but then you would have to deal with overlapping ranges once. I think you could use SpectraMerging.createNewNonOverlappingRange() for it

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java [hunk 2]
```diff
@@ -166,6 +134,52 @@ private List<FeatureListRow> generatePseudoSpectraByRtOnly(List<ModularFeature>
...
+      }
+      if (!isIgnored) {
+        return feature;
+      }
+    }
+    return null; // Return null if all features are in the ignored ranges
+  }
+
+  private void createNewDeconvolutedFeatureList(List<FeatureListRow> deconvolutedFeatureListRows) {
+    deconvolutedFeatureList = FeatureListUtils.createCopy(featureList, suffix,
```
Comment (SteffenHeu) [inline mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java:171]: dont forget to add the applied methods

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+          mzs, intensities, mainFeature.getRepresentativeScan().getPolarity(),
+          "Correlated Features Pseudo Spectrum", PseudoSpectrumType.GC_EI);
+
+      mainFeature.setAllMS2FragmentScans(List.of(pseudoSpectrum));
+      deconvolutedFeatureListRowsByRtOnly.add(mainFeature.getRow());
+    }
+    return deconvolutedFeatureListRowsByRtOnly;
+  }
+
+  public static ModularFeature getMainFeature(List<ModularFeature> groups,
```
Comment (SteffenHeu) [inline mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java:72]: I feel like i've just seen this method :D

Comment (SteffenHeu) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java:72]: do the features need to be sored in some way for this method to be reproducible? e.g by descending hight? if yes add to javadoc

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import io.github.mzmine.datamodel.features.ModularFeature;
+import io.github.mzmine.datamodel.impl.SimplePseudoSpectrum;
+import io.github.mzmine.modules.MZmineProcessingStep;
+import io.github.mzmine.parameters.ParameterSet;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+
+public class SpectralDeconvolutionTools {
```
Comment (SteffenHeu) [inline mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java:42]: call utils to adhere to naming scheme

**Review by SteffenHeu — COMMENTED** [review]
Looks good, i got some comments though

File: mzmine-community/src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYScatterPlot.java [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+ * Generic plot class that can be used to plot everything that implements the
+ * {@link PlotXYDataProvider} interface or is a {@link ColoredXYDataset}.
+ *
+ * @param <T>
+ * @author https://github.com/SteffenHeu
+ * @see ExampleXYProvider
+ * @see ColoredXYDataset
+ * @see PlotXYDataProvider
+ */
+public class SimpleXYScatterPlot<T extends PlotXYDataProvider> extends EChartViewer implements
```
[the remark this answers is not in the document]
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYScatterPlot.java:90]: I tried using the SimpleXYScatterPlot. I ran into issues setting the dataset colors. If you know how to transition to SimpleXYScatterPlot feel free to do it.
There is also room for improvement regarding the renderer. Would be cool if every data point has the width of the rt range

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java (hunk 3 above)
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionTools.java:72]: yes, moved the method to the tools class and forgot to remove it.
I added the java doc

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java (hunk 2 above)
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCTask.java:106]: done in createCopy

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+  }
+
+  private HBox buildScatterPlotLegend() {
+    HBox legendBox = new HBox(20);
+
+    // Create the legend item for the main feature RT range
+    Label mainFeatureRtRangeLbl = new Label("Compound rt range");
+    Rectangle mainFeatureRtRangeRect = new Rectangle(20, 20,
+        javafx.scene.paint.Color.rgb(200, 200, 255, 0.39));
+    HBox mainFeatureLegendItem = new HBox(5, mainFeatureRtRangeRect, mainFeatureRtRangeLbl);
```
[the remark this answers is not in the document]
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java:151]: cool

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    scatterPlotLegend = buildScatterPlotLegend();
+    scatterPlotBorderPane.setBottom(scatterPlotLegend);
+    clusteringSelectedFeatureSplit = new SplitPane();
+    clusteringSelectedFeatureSplit.getItems().add(scatterPlotBorderPane);
+    clusteringSelectedFeatureSplit.setOrientation(Orientation.VERTICAL);
+    previewWrapperPane.setCenter(clusteringSelectedFeatureSplit);
+
+    deconvolutedFeaturesComboBox = new ComboComponent<>(FXCollections.observableArrayList());
+    HBox buttonBox = new HBox(updateButton, deconvolutedFeaturesComboBox);
+    deconvolutedFeaturesComboBox.setOnAction(_ -> updateSelectedFeature());
```
[the remark this answers is not in the document]
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java:122]: looks good

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+    double rtDiff = feature.getRT() - rtValue;
+    double mzDiff = feature.getMZ() - mzValue;
+    return Math.sqrt(rtDiff * rtDiff + mzDiff * mzDiff);
+  }
+
+  private void updateSelectedFeature() {
+    ModularFeature selectedFeature = (ModularFeature) deconvolutedFeaturesComboBox.getSelectionModel()
+        .getSelectedItem();
+    if (selectedFeature != null) {
+      pseudoSpectrumVisualizerPane = new PseudoSpectrumVisualizerPane(selectedFeature);
```
[the remark this answers is not in the document]
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java:220]: no

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    Feature closestFeature = allFeatures.stream().min(Comparator.comparingDouble(
+            feature -> calculateEuclideanDistance(feature, rtValue, mzValue)))
+        .orElse(null);
+
+    if (closestFeature != null) {
+      for (List<ModularFeature> features : groupedFeatures) {
+        for (ModularFeature feature : features) {
+          if (feature.equals(closestFeature)) {
+            return SpectralDeconvolutionTools.getMainFeature(features, mzValuesToIgnore);
+          }
```
[the remark this answers is not in the document]
Comment (ansgarkorf) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_spectraldeconvolutiongc/SpectralDeconvolutionGCDialog.java]: not sure whart to do here

**Comment by SteffenHeu** [comment]
looks good, thanks

**Pull request merged**